src/main/scala/firrtl/transforms/GroupComponents.scala

```diff
@@ -0,0 +1,340 @@
+          Block(topStmts)
+        // TODO Attach if all are in a group?
+        case _: IsDeclaration|_: Connect|_: Attach =>
+          // Sink is in Top
+          val ret = s mapExpr inTopFixExps
+          ret
+        case other => other map onStmt
+      }
+    }
+
+
+    // Build datastructures
+    val newTopBody = Block(labelOrder.map(g => WDefInstance(NoInfo, label2instance(g), label2module(g), UnknownType)) ++ Seq(onStmt(m.body)))
+    val finalTopBody = Block(Utils.squashEmpty(newTopBody).asInstanceOf[Block].stmts.distinct)
+
+    val newModules = labelOrder.filter(_ != "") map { group =>
+      Module(NoInfo, label2module(group), groupPorts(group).distinct, Block(groupStatements(group).distinct))
+    }
+    Seq(m.copy(body = finalTopBody)) ++ newModules
+  }
+
+  def getWRef(e: Expression): WRef = e match {
+    case w: WRef => w
+    case other =>
+      var w = WRef("")
+      other mapExpr { e => w = getWRef(e); e}
+      w
+  }
+
+  /**
+    * Compute how each component connects to each other component
+    * It is non-directioned; there is an edge from source to sink and from sink to souce
+    * @param m module to compute connectivity
+    * @return a bi-directional representation of component connectivity
+    */
+  def getComponentConnectivity(m: Module): MutableDiGraph[String] = {
+    val bidirGraph = new MutableDiGraph[String]
+    val simNamespace = Namespace()
+    val simulations = new mutable.HashMap[String, Statement]
+    def onExpr(sink: WRef)(e: Expression): Expression = e match {
+      case w@WRef(name, _, _, _) =>
+        bidirGraph.addPairWithEdge(sink.name, name)
+        bidirGraph.addPairWithEdge(name, sink.name)
+        w
+      case other => other map onExpr(sink)
+    }
+    def onStmt(stmt: Statement): Unit = stmt match {
+      case w: WDefInstance =>
+      case h: IsDeclaration => h map onExpr(WRef(h.name))
+      case Attach(_, exprs) => // Add edge between each expression
+        exprs.tail map onExpr(getWRef(exprs.head))
+      case Connect(_, loc, expr) =>
+        onExpr(getWRef(loc))(expr)
+      case q@Stop(_,_, clk, en) =>
+        val simName = simNamespace.newTemp
+        simulations(simName) = q
+        Seq(clk, en) map onExpr(WRef(simName))
+      case q@Print(_, _, args, clk, en) =>
+        val simName = simNamespace.newTemp
+        simulations(simName) = q
+        (args :+ clk :+ en) map onExpr(WRef(simName))
+      case Block(stmts) => stmts.foreach(onStmt)
+      case ignore @ (_: IsInvalid | EmptyStmt) => // do nothing
+      case other => throw new Exception(s"Unexpected Statement $other")
+    }
+
+    onStmt(m.body)
+    m.ports.foreach { p =>
+      bidirGraph.addPairWithEdge("", p.name)
+      bidirGraph.addPairWithEdge(p.name, "")
+    }
+    bidirGraph
+  }
+}
+
+/**
+  * Splits a module into multiple modules by grouping its components via [[GroupAnnotation]]'s
+  * Tries to deduplicate the resulting circuit
+  */
+class GroupAndDedup extends GroupComponents {
```

Comment:
Rather than this inheriting from GroupComponents, I think it should just be a different Transform that calls GroupComponents.